File: src/main/java/cubesrv/CubeCosets.java
package cubesrv;

import static cubesrv.CubeDefs.*;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class CubeCosets {

    public static class CubeCosetsAtDepth {
        @SuppressWarnings("unchecked")
        final Map<cubeedges, List<cube>>[] m_itemsArr = new HashMap[2187];

        public CubeCosetsAtDepth() {
            for (int i = 0; i < 2187; i++) m_itemsArr[i] = new HashMap<>();
        }

        public boolean addCube(int ccoReprIdx, cubeedges ceRepr, cube c) {
            Map<cubeedges, List<cube>> items = m_itemsArr[ccoReprIdx];
            List<cube> cubeList = items.computeIfAbsent(ceRepr, k -> new ArrayList<>());
            boolean res = cubeList.isEmpty();
            cubeList.add(c);
            return res;
        }

        public boolean containsCCOrients(int ccoReprIdx) {
            return !m_itemsArr[ccoReprIdx].isEmpty();
        }

        public List<cube> getCubesForCE(int ccoReprIdx, cubeedges ceRepr) {
            return m_itemsArr[ccoReprIdx].get(ceRepr);
        }
    }

    private final CubeCosetsAtDepth[] m_cubesAtDepths;
    private int m_availCount = 0;

    public CubeCosets(int size) {
        m_cubesAtDepths = new CubeCosetsAtDepth[size];
        for (int i = 0; i < size; i++) m_cubesAtDepths[i] = new CubeCosetsAtDepth();
    }

    public int availCount()    { return m_availCount; }
    public void incAvailCount(){ ++m_availCount; }
    public int availMaxCount() { return m_cubesAtDepths.length; }
    public CubeCosetsAtDepth getAt(int idx) { return m_cubesAtDepths[idx]; }
}

File: src/main/kotlin/cubesrv/CubeCosets.kt
package cubesrv

/* A set of spaces (cosets of BG space) reachable at specific depth.
 * A coset is identified by a representative cube.
 */
class CubeCosetsAtDepth {
    // Map<edges of representative cube, cubes reachable at depth>
    // the representative cube corners permutation is identity; 2^7 corner orientations
    val m_itemsArr = Array<MutableMap<cubeedges, MutableList<cube>>>(2187) {
        mutableMapOf<cubeedges, MutableList<cube>>() }

    /* Adds a cube to coset identified by corner orients and representative cubeedges.
     */
    fun addCube(ccoReprIdx : Int, ceRepr : cubeedges, c : cube) : Boolean {
        val items : MutableMap<cubeedges, MutableList<cube>> = m_itemsArr[ccoReprIdx]
        val cubeList : MutableList<cube> = items.getOrPut(ceRepr) { mutableListOf<cube>() }
        val res : Boolean = cubeList.isEmpty()
        cubeList.add(c)
        return res
    }

    fun containsCCOrients(ccoReprIdx : Int) : Boolean {
        return !m_itemsArr[ccoReprIdx].isEmpty()
    }

    fun getCubesForCE(ccoReprIdx : Int, ceRepr : cubeedges) : List<cube>?  {
        val items : MutableMap<cubeedges, MutableList<cube>> = m_itemsArr[ccoReprIdx]
        return items.get(ceRepr)
    }
}

class CubeCosets(size : Int) {
    val m_cubesAtDepths = Array<CubeCosetsAtDepth>(size) { CubeCosetsAtDepth(); }
    var m_availCount : Int = 0

    fun availCount() : Int = m_availCount
    fun incAvailCount() { ++m_availCount; }
    fun availMaxCount() : Int = m_cubesAtDepths.size
    fun getAt(idx : Int) = m_cubesAtDepths[idx]
}
